src/x86.rs

```diff
@@ -132,222 +145,216 @@ impl<V: Vector> VectorHash<V> {
-        ///
-        /// In the case where the needle is not a multiple of the number of SIMD
-        /// lanes, the last chunk is made up of a partial overlap with the
-        /// penultimate chunk to avoid reading random memory, differing from the
-        /// original implementation. In this case, a mask is used to prevent
-        /// performing an equality comparison on the same position twice.
-        ///
-        /// When the haystack is too short for an AVX2 register, a similar SSE2
-        /// fallback is used instead. Finally, for very short haystacks there is
-        /// a scalar Rabin-Karp implementation.
-        pub struct $name {
-            needle: Box<[u8]>,
-            position: usize,
-            scalar_hash: ScalarHash,
-            sse2_hash: VectorHash<__m128i>,
-            avx2_hash: VectorHash<__m256i>,
+/// Single-substring searcher using an AVX2 algorithm based on the
+/// "Generic SIMD" algorithm [presented by Wojciech
+/// Muła](http://0x80.pl/articles/simd-strfind.html).
+///
+/// It is similar to the Rabin-Karp algorithm, except that the hash is
+/// not rolling and is calculated for several lanes at once. It begins
+/// by picking the first byte in the needle and checking at which
+/// positions in the haystack it occurs. Any position where it does not
+/// can be immediately discounted as a potential match.
+///
+/// We then repeat this idea with a second byte in the needle (where the
+/// haystack is suitably offset) and take a bitwise AND to further limit
+/// the possible positions the needle can match in. Any remaining
+/// positions are fully evaluated using an equality comparison with the
+/// needle.
+///
+/// Originally, the algorithm always used the last byte for this second
+/// byte. Whilst this is often the most efficient option, it is
+/// vulnerable to a worst-case attack and so this implementation instead
+/// allows any byte (including a random one) to be chosen.
+///
+/// In the case where the needle is not a multiple of the number of SIMD
+/// lanes, the last chunk is made up of a partial overlap with the
+/// penultimate chunk to avoid reading random memory, differing from the
+/// original implementation. In this case, a mask is used to prevent
+/// performing an equality comparison on the same position twice.
+///
+/// When the haystack is too short for an AVX2 register, a similar SSE2
+/// fallback is used instead. Finally, for very short haystacks there is
+/// a scalar Rabin-Karp implementation.
+pub struct Avx2Searcher<N: Needle> {
+    position: usize,
+    scalar_hash: ScalarHash,
+    sse2_hash: VectorHash<__m128i>,
+    avx2_hash: VectorHash<__m256i>,
+    needle: N,
+}
+
+impl<N: Needle> Avx2Searcher<N> {
+    /// Creates a new searcher for `needle`. By default, `position` is
+    /// set to the last character in the needle.
+    #[target_feature(enable = "avx2")]
+    pub unsafe fn new(needle: N) -> Self {
+        let position = needle.as_bytes().len() - 1;
```

Comment:
Theoretically we should now use `needle.size()` here (same goes for the two length checks `with_position()`) although I imagine the compiler has probably already propagated the constant.

Reply:
For this line I think it makes sense to use `needle.size()` but in `with_position` I need to have the needle as bytes anyway so I am not sure it really matters.

Reply:
Cool, that makes sense